src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -148,6 +154,44 @@ public JsonClassInfo(Type type, JsonSerializerOptions options)
                                     // Non-public properties should not be included for (de)serialization.
                                 }
                             }
+
+                            foreach (FieldInfo fieldInfo in currentType.GetFields(binding))
```

Comment:
Any way to extract the following logic into a method and share between fields and properties above?

Reply:
It's totally possible to have only loops invoking a generic method accepting two non-allocating delegates or usual delegates to check visibility and to create a property.

Reply:
I was thinking about something along these lines to avoid some of the code duplication. Note, I included the checks for whether we should actually include the field:

```cs
for (Type? currentType = type; currentType != null; currentType = currentType.BaseType)
{
    const BindingFlags bindingFlags =
        BindingFlags.Instance |
        BindingFlags.Public |
        BindingFlags.NonPublic |
        BindingFlags.DeclaredOnly;

    foreach (PropertyInfo propertyInfo in currentType.GetProperties(bindingFlags))
    {
        // Ignore indexers
        if (propertyInfo.GetIndexParameters().Length > 0)
        {
            continue;
        }

        bool isPublic = propertyInfo.GetMethod?.IsPublic == true || propertyInfo.SetMethod?.IsPublic == true;
        if (isPublic)
        {
            CacheMember(propertyInfo, propertyInfo.PropertyType, currentType, cache, options);
        }
        else
        {
            if (JsonPropertyInfo.GetAttribute<JsonIncludeAttribute>(propertyInfo) != null)
            {
                ThrowHelper.ThrowInvalidOperationException_JsonIncludeOnNonPublicInvalid(propertyInfo, currentType);
            }

            // Non-public properties should not be included for (de)serialization.
        }
    }

    foreach (FieldInfo fieldInfo in currentType.GetFields(bindingFlags))
    {
        bool hasJsonInclude = JsonPropertyInfo.GetAttribute<JsonIncludeAttribute>(fieldInfo) != null;
        bool shouldInclude = options.IncludeFields || hasJsonInclude;

        if (!fieldInfo.IsPublic)
        {
            if (hasJsonInclude)
            {
                ThrowHelper.ThrowInvalidOperationException_JsonIncludeOnNonPublicInvalid(fieldInfo, currentType);
            }

            // Non-public fields should not be included for (de)serialization.
        }
        else if (shouldInclude)
        {
            CacheMember(fieldInfo, fieldInfo.FieldType, currentType, cache, options);
        }
    }
}

private void CacheMember(
    MemberInfo memberInfo,
    Type memberType,
    Type parentClassType,
    Dictionary<string, JsonPropertyInfo> cache,
    JsonSerializerOptions options)
{
    JsonPropertyInfo jsonPropertyInfo = AddProperty(memberInfo, memberType, parentClassType, options);
    Debug.Assert(jsonPropertyInfo.NameAsString != null);

    // If the JsonPropertyNameAttribute or naming policy results in collisions, throw an exception.
    if (!JsonHelpers.TryAdd(cache, jsonPropertyInfo.NameAsString, jsonPropertyInfo))
    {
        JsonPropertyInfo other = cache[jsonPropertyInfo.NameAsString];

        if (other.ShouldDeserialize == false && other.ShouldSerialize == false)
        {
            // Overwrite the one just added since it has [JsonIgnore].
            cache[jsonPropertyInfo.NameAsString] = jsonPropertyInfo;
        }
        else if (other.MemberInfo?.Name != jsonPropertyInfo.MemberInfo?.Name &&
            (jsonPropertyInfo.ShouldDeserialize == true || jsonPropertyInfo.ShouldSerialize == true))
        {
            // Check for name equality is required to determine when a new slot is used for the member.
            // Therefore, if names are not the same, there is conflict due to the name policy or attributes.
            ThrowHelper.ThrowInvalidOperationException_SerializerPropertyNameConflict(Type, jsonPropertyInfo);
        }
        // else ignore jsonPropertyInfo since it has [JsonIgnore] or it's hidden by a new slot.
    }
}
```